File: frontend/src/types/trading.ts
import type { PageResult } from './api';

export interface ManualOrderRequest {
  symbol: string;
  name?: string | null;
  side: 'BUY' | 'SELL';
  price: number;
  quantity: number;
  order_type?: string;
}

export interface SignalOrderRequest {
  price?: number | null;
  quantity?: number | null;
}

export interface TradingPosition {
  id: number;
  account_id: string;
  symbol: string;
  name: string;
  quantity: number;
  available_quantity: number;
  cost_price: number;
  last_price: number;
  market_value: number;
  pnl: number;
  pnl_ratio: number;
  snapshot_time: string;
}

export interface TradingOrderRecord {
  id: number;
  local_order_id: string;
  qmt_order_id?: string | null;
  account_id: string;
  symbol: string;
  name: string;
  side: string;
  price: number;
  quantity: number;
  filled_quantity: number;
  status: string;
  qmt_status?: string | null;
  source: string;
  strategy_id?: string | null;
  strategy_name?: string | null;
  signal_id?: string | null;
  idempotency_key?: string | null;
  order_time: string;
  updated_at: string;
}

export interface TradingTradeRecord {
  id: number;
  trade_id: string;
  local_order_id?: string | null;
  qmt_order_id?: string | null;
  account_id: string;
  symbol: string;
  name: string;
  side: string;
  price: number;
  quantity: number;
  amount: number;
  fee: number;
  source: string;
  strategy_name?: string | null;
  trade_time: string;
}

export interface TradingSignalRecord {
  id: number;
  strategy_id: number;
  strategy_name: string;
  run_id: string;
  symbol: string;
  name: string;
  action: string;
  price: number;
  amount?: number | null;
  reason: string;
  status: string;
  signal_time: string;
  order_id?: string | null;
  created_at: string;
}

export interface ExecutionLogRecord {
  id: number;
  local_order_id?: string | null;
  level: string;
  message: string;
  technical_detail?: string | null;
  created_at: string;
}

export interface OrderSubmitResult {
  order: TradingOrderRecord;
  message: string;
  duplicate: boolean;
}

export type TradingPositionPage = PageResult<TradingPosition>;
export type TradingOrderPage = PageResult<TradingOrderRecord>;
export type TradingTradePage = PageResult<TradingTradeRecord>;
export type TradingSignalPage = PageResult<TradingSignalRecord>;
export type ExecutionLogPage = PageResult<ExecutionLogRecord>;

File: backend/schemas/trading.py
from pydantic import BaseModel, Field, field_validator

from backend.schemas.common import PageResult


class ManualOrderRequest(BaseModel):
    symbol: str = Field(min_length=1)
    name: str | None = None
    side: str
    price: float = Field(gt=0)
    quantity: int = Field(gt=0)
    order_type: str = "限价委托"


class SignalOrderRequest(BaseModel):
    price: float | None = Field(default=None, gt=0)
    quantity: int | None = Field(default=None, gt=0)


class TradingPosition(BaseModel):
    id: int
    account_id: str
    symbol: str
    name: str
    quantity: int
    available_quantity: int
    cost_price: float
    last_price: float
    market_value: float
    pnl: float
    pnl_ratio: float
    snapshot_time: str


class TradingOrderRecord(BaseModel):
    id: int
    local_order_id: str
    qmt_order_id: str | None = None
    account_id: str
    symbol: str
    name: str
    side: str
    price: float
    quantity: int
    filled_quantity: int
    status: str
    qmt_status: str | None = None
    source: str
    strategy_id: str | None = None
    strategy_name: str | None = None
    signal_id: str | None = None
    idempotency_key: str | None = None
    order_time: str
    updated_at: str

    @field_validator("status", mode="before")
    @classmethod
    def normalize_status(cls, value: object) -> str:
        status = str(value or "").strip()
        return "待同步" if status in {"", "未知", "unknown", "None"} else status


class TradingTradeRecord(BaseModel):
    id: int
    trade_id: str
    local_order_id: str | None = None
    qmt_order_id: str | None = None
    account_id: str
    symbol: str
    name: str
    side: str
    price: float
    quantity: int
    amount: float
    fee: float
    source: str
    strategy_name: str | None = None
    trade_time: str


class TradingSignalRecord(BaseModel):
    id: int
    strategy_id: int
    strategy_name: str
    run_id: str
    symbol: str
    name: str
    action: str
    price: float
    amount: float | None = None
    reason: str
    status: str
    signal_time: str
    order_id: str | None = None
    created_at: str


class ExecutionLogRecord(BaseModel):
    id: int
    local_order_id: str | None = None
    level: str
    message: str
    technical_detail: str | None = None
    created_at: str


class OrderSubmitResult(BaseModel):
    order: TradingOrderRecord
    message: str
    duplicate: bool = False


TradingOrderPage = PageResult[TradingOrderRecord]
TradingTradePage = PageResult[TradingTradeRecord]
TradingPositionPage = PageResult[TradingPosition]
TradingSignalPage = PageResult[TradingSignalRecord]
ExecutionLogPage = PageResult[ExecutionLogRecord]
